src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -1104,154 +1137,280 @@ TargetShardIntervalForModify(Oid distributedTableId, Query *query,
 		{
-			targetCountType = "multiple";
+			if (route->shardId == shardId)
+			{
+				foundSameShardId = true;
+			}
+			else
+			{
+				/* new shard id seen; current aggregation done; add to list */
+				modifyRouteList = lappend(modifyRouteList, route);
+			}
 		}
 
-		if (prunedShardCount == 0)
+		if (foundSameShardId)
 		{
-			appendStringInfo(errorHint, "Make sure you have created a shard which "
-										"can receive this partition column value.");
+			/*
+			 * Our current value has the same shard id as our aggregate object,
+			 * so append the rowValues.
+			 */
+			route->rowValuesLists = lappend(route->rowValuesLists,
+											insertValues->rowValues);
 		}
 		else
 		{
-			appendStringInfo(errorHint, "Make sure the value for partition column "
-										"\"%s\" falls into a single shard.",
-							 partitionColumnName);
+			/* we encountered a new shard id; build a new aggregate object */
+			route = (ModifyRoute *) palloc(sizeof(ModifyRoute));
+			route->shardId = insertValues->shardId;
+			route->partitionvarno = partitionvarno;
+			route->rowValuesLists = list_make1(insertValues->rowValues);
 		}
+	}
 
-		appendStringInfo(errorMessage, "cannot run INSERT command which targets %s "
-									   "shards", targetCountType);
+	/* left holding one final aggregate object; add to list */
+	modifyRouteList = lappend(modifyRouteList, route);
 
-		(*planningError) = DeferredError(ERRCODE_FEATURE_NOT_SUPPORTED,
-										 errorMessage->data, NULL,
-										 errorHint->data);
+	return modifyRouteList;
+}
 
-		return NULL;
-	}
 
-	return (ShardInterval *) linitial(prunedShardList);
+/*
+ * CompareInsertValuesByShardId does what it says in the name. Used for sorting
+ * InsertValues objects by their shard.
+ */
+static int
+CompareInsertValuesByShardId(const void *leftElement, const void *rightElement)
+{
+	InsertValues *leftValue = *((InsertValues **) leftElement);
+	InsertValues *rightValue = *((InsertValues **) rightElement);
+	int64 leftShardId = leftValue->shardId;
+	int64 rightShardId = rightValue->shardId;
+
+	/* we compare 64-bit integers, instead of casting their difference to int */
+	if (leftShardId > rightShardId)
+	{
+		return 1;
+	}
+	else if (leftShardId < rightShardId)
+	{
+		return -1;
+	}
+	else
+	{
+		return 0;
+	}
 }
 
 
 /*
- * FindShardForUpdateOrDelete finds the shard interval in which an UPDATE or
- * DELETE command should be applied, or sets planningError when the query
- * needs to be applied to multiple or no shards.
+ * BuildRouteForUpdateOrDelete finds the shard interval in which an UPDATE or
+ * DELETE command should be applied and builds a ModifyRoute object for routing
+ * the command. If the query involves multiple or no shards, the function will
+ * set planningError instead.
  */
-static ShardInterval *
-FindShardForUpdateOrDelete(Query *query, DeferredErrorMessage **planningError)
+static ModifyRoute *
+BuildRouteForUpdateOrDelete(Query *query, DeferredErrorMessage **planningError)
```

Comment:
Yeah it was mostly a consistency thing. I didn't want the two paths (`INSERT` vs `UPDATE`/`DELETE`) to have different return types. It only returns one route right now but it makes the rest of the code flow better. Is there anything you wanted me to do here in the code?